src/RFC-0325_DanTimeManagement.md

```diff
@@ -0,0 +1,188 @@
+
+## DAN Lag
+
+Because the [Minotari] chain influences the DAN layer and is used as a timing mechanism for the DAN, we need to ensure
+that the DAN has a stable view of the [Minotari].
+
+The [Minotari] is built on Proof-of-Work and this means that chain might undergo re-organisation. We need to ensure
+that re-orgs do cause a [VNC] reshuffle. We introduce a concept called `DAN Lag` which is an offset between when a
+[VNC] change is recorded on the Minotari chain and when it takes effect.
+
+For example, if we define `DAN Lag` as 720 blocks, or a day. Then when a VN registers and that transaction is
+mined in the [Minotari] at height 1000 then only at height 1720 will the DAN Layer recognise the [VN] as being
+registered.
+
+This `DAN Lag` allows the [Minotari] chain to have small re-orgs of less than the `DAN Lag` without it having any
+effect on the DAN Layer.
+An added benefit of this is that the DAN Layer knows all changes in advance of when it will happen, so when a [VN] is
+swapped to a new [shard space] it will give it the `DAN Lag` period to sync the required state.
+
+## DAN Grace Time
+
+Minotari nodes are decentralized. Thus, we don't have a single point of view of the state of the chain.
+
+The DAN operates in ms timeframes, while the [Minotari] chain runs at the minute timeframe, averaging 2 minutes per
+block. These blocks also take a few seconds to process and propagate through the network.
+
+This means that for some [VN]s the [Minotari] block height might be 999, while for others it might be 1000. The
+practical problem for this is that we cannot base consensus decisions on the block height if the [VN]s cannot come to
+some consensus as to what the [Minotari] height is.
+
+We define `DAN Grace Time` or DGT as the number of blocks in the past or future of the [VN]'s own current block height,
+that it will accept. This is very similar to the [FTL] concept concerning the timestamp in the block header.
+
+In example of this would be, that we have `VN_1` whose [base node] is on height 999 and `VN_2` whose [base node] is on
+height 1000. We have set the DGT as 1.
+An instruction has specified that from height 1000 the [VNC] that must process it contain both `VN_1` and `VN_2`, but
+before that it is only `VN_1`. From this it can be seen that `VN_1` thinks only it must be in the [VNC] but `VN_2`
+thinks they both need to be. But using the DGT of 1 block, `VN_1` will accept `VN_2` as part of the [VNC]
+because it is withing the DGT period and it assumes that its own [base node] is simply lagging 1 block.
+
+The DGT has the additional benefit of allowing a [VN] to finish processing an instruction after a block height has been
+reached.
+
+With the example mentioned above if `VN_1` has to stop processing an instruction at height 1000, but it started the
+instruction at height 999. When height 1000 comes
+along it can still continue to process the instruction till height 1001.
```

Comment:
I feel that a more effective solution to this problem is to treat this as a consensus issue on the DAN layer.

A leader can propose `NEW_EPOCH` to the presumed new VNC. If less than 66% of VNs have updated their epoch, the motion fails, and the VNC cannot form.

We should do a rough safety and liveness proof of whether this works based on:
- an honest node does not equivocate on its epoch. Once a `NEW_EPOCH` is accepted, a VN will ignore messages from their old VNC/epoch combination.
- A VNC must respond to and accept valid proposals from the old VNC/epoch combo until a new epoch has been accepted.

One consequence of this is that we may need to shuffle 33%-66% of nodes every epoch so that consensus and a new committee lines up in terms of numbers.